packages/yarnpkg-fslib/sources/algorithms/watchFile.ts

```diff
@@ -1,167 +1,86 @@
-    this.path = path;
-    this.bigint = bigint;
-
-    this.lastStats = this.stat();
-  }
-
-  start() {
-    assertStatus(this.status, Status.Ready);
-    this.status = Status.Running;
-
-    // Node allows other listeners to be registered up to 3 milliseconds
-    // after the watcher has been started, so that's what we're doing too
-    setTimeout(() => {
-      // Per the Node FS docs:
-      // "When an fs.watchFile operation results in an ENOENT error,
-      // it will invoke the listener once, with all the fields zeroed
-      // (or, for dates, the Unix Epoch)."
-      if (!this.fakeFs.existsSync(this.path)) {
-        this.emit(Event.Change, this.lastStats, this.lastStats);
-      }
-    }, 3);
+import {FakeFS, WatchFileOptions, WatchFileCallback} from '../FakeFS';
+import {Path}                                        from '../path';
+
+import {CustomStatWatcher}                           from './watchFile/CustomStatWatcher';
+
+const statWatchersByFakeFS = new WeakMap<FakeFS<Path>, Map<Path, CustomStatWatcher<Path>>>();
+
+export function watchFile<P extends Path>(
+  fakeFs: FakeFS<P>,
+  path: P,
+  a: WatchFileOptions | WatchFileCallback,
+  b?: WatchFileCallback
+) {
+  let bigint: boolean;
+  let persistent: boolean;
+  let interval: number;
+
+  let listener: WatchFileCallback;
+
+  switch (typeof a) {
+    case `function`: {
+      bigint = false;
+      persistent = true;
+      interval = 5007;
+
+      listener = a;
+    } break;
+
+    default: {
+      ({
+        bigint = false,
+        persistent = true,
+        interval = 5007,
+      } = a);
+
+      listener = b!;
+    } break;
   }
 
-  stop() {
-    assertStatus(this.status, Status.Running);
-    this.status = Status.Stopped;
-
-    this.emit(Event.Stop);
-  }
-
-  stat() {
-    try {
-      return this.fakeFs.statSync(this.path);
-    } catch (error) {
-      if (error.code === `ENOENT`) {
-        return statUtils.makeEmptyStats();
-      } else {
-        throw error;
-      }
-    }
-  }
-
-  /**
-   * Creates an interval whose callback compares the current stats with the previous stats and notifies all listeners in case of changes.
-   *
-   * @param opts.persistent Decides whether the interval should be immediately unref-ed.
-   */
-  makeInterval(opts: ListenerOptions) {
-    const interval = setInterval(() => {
-      const currentStats = this.stat();
-      const previousStats = this.lastStats;
-
-      if (statUtils.areStatsEqual(currentStats, previousStats))
-        return;
+  let statWatchers = statWatchersByFakeFS.get(fakeFs);
+  if (typeof statWatchers === `undefined`)
+    statWatchersByFakeFS.set(fakeFs, statWatchers = new Map());
 
-      this.lastStats = currentStats;
+  let statWatcher = statWatchers.get(path);
+  if (typeof statWatcher === `undefined`) {
+    statWatcher = new CustomStatWatcher<P>(fakeFs, path, {bigint});
 
-      this.emit(Event.Change, currentStats, previousStats);
-    }, opts.interval);
+    statWatcher.start();
```

Comment:
nit: Maybe put both of these together in a single call? If you don't want to do the setTimeout in the constructor, that's fair, but you can mark the constructor private and declare a public static `create` function that would create the watcher + start it.